src/smtp_observer.py

```diff
@@ -0,0 +1,162 @@
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "smtp-observer")
+        self._charm = charm
+        self.smtp = SmtpRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+        )
+        self.framework.observe(
+            self.smtp.on.smtp_data_available,
+            self._on_smtp_relation_data_available,
+        )
+
+    def get_relation_as_smtp_conf(self) -> Optional[SMTPConfiguration]:
+        """Get SMTP data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+        """
+        try:
+            relation_data: Optional[SmtpRelationData] = self.smtp.get_relation_data()
+        except ValidationError:
+            # ValidationError happens in the smtp(_legacy)relation_created event, as
+            # the relation databag is empty at that point.
+            logger.info("SMTP databag is empty. SMTP information will be set in the next event.")
+            return None
+
+        if relation_data is None:
+            return None
+
+        user = None
+        password = None
+        if relation_data.auth_type == AuthType.PLAIN:
+            user = relation_data.user
+            password = self._get_password_from_relation_data(relation_data)
+
+        # Not all combinations for the next variables are correct. See:
+        # https://github.com/matrix-org/synapse/blob/develop/synapse/config/emailconfig.py
+        force_tls = False
+        enable_tls = False
+        require_transport_security = False
+        if relation_data.transport_security == TransportSecurity.STARTTLS:
+            enable_tls = True
+            require_transport_security = True
+        elif relation_data.transport_security == TransportSecurity.TLS:
+            force_tls = True
+            enable_tls = True
+            require_transport_security = True
+
+        return SMTPConfiguration(
+            enable_tls=enable_tls,
+            force_tls=force_tls,
+            require_transport_security=require_transport_security,
+            host=relation_data.host,
+            port=relation_data.port,
+            user=user,
+            password=password,
+        )
+
+    def _get_password_from_relation_data(self, relation_data: SmtpRelationData) -> Optional[str]:
+        """Get smtp password from relation data.
+
+        Arguments:
+            relation_data: The relation data from where to extract the password
+
+        Returns:
+            the password or None if no password found
+        """
+        # If the relation data password_id exists, that means that
+        # Juju version is >= 3.0 and secrets are used for the password.
+        # Otherwise, use the field password as a fallback
+        if relation_data.password_id:
+            secret = self.model.get_secret(id=relation_data.password_id)
```

Comment:
I'm considering what would happen if this raises SecretNotFound error. Since there is no action to be taken, we could let this put the charm in an error state if it happens.

Reply:
Not really sure. Could it happen that there is a relation but the owner (smtp integrator charm)  deletes the secret? If that happens, would just be an option to put the charm in error state until the integration is removed?

Reply:
I think in this case makes more sense to let the Exception raise and create an error.